Import sd-gpg key from  dom0 on boot

deeplow opened the pull request:
Fixes https://github.com/freedomofpress/securedrop-workstation/issues/1505 by importing sd-gpg key on boot. We can't make it yet a disposable, because in some cases multiple keys could be present in `sd-gpg` and those are not backed up in dom0.

NOTE: I named the RPC policy as `keys`  and not `key` because in the future we'll probably want to add support for multiple ones.

<strike>Also [adds a mechanism to run scripts on boot](https://github.com/freedomofpress/securedrop-workstation/pull/1521/files#diff-296179b3a998617cd803e9293c9bb52dd4e748fae47065b28b74ad17ec4e34c3R42-R45) (it needs a counterpart component in the client repo). Some other alternatives being documented in https://github.com/freedomofpress/securedrop-workstation/issues/1522.</strike>

## Test plan
<!-- Delete this section if not applicable (e.g., some docs-only changes) -->
Setup:
- [ ] Do not delete your existing `sd-gpg`
- [ ] `./scripts/try-client-pr.py 3096`

Testing:
  - [ ] run `make dev`
- [x] open a terminal in `sd-gpg` and:
  - [x] check that the key is there (`gpg -K` should list it)
  - [x] `env | grep QUBES_GPG_AUTOACCEPT` should list it as set to `2147483647`
  - [x] `tail /home/user/.profile` — it should not have any `QUBES_GPG_AUTOACCEPT=...`
  - [x] remove the key you have there (either nuking the keyring `rm /home/user/.gnupg/pubring.kbx` or removing it with `gpg`)
- [x] Restart `sd-gpg` and open a terminal there:
  - [x] check that the key is there (`gpg -K` should list it)

## Checklist

<!-- If you leave any box below unchecked, please clarify where you may need support.
     If you're unsure, that's fine — a reviewer can help you out. -->

This change accounts for:
- [ ] any necessary RPM packaging updates (e.g., added/removed files, see `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`)
- [ ] any required documentation

deeplow committed:
Allow sd-gpg to obtain secret key in dom0 via qrexec

This implements a minimal Qubes RPC service that simply cats the secret
key file in dom0. Since it only cats a regular file and takes in no
parameters, it is security-wise. This serves as a replacement for
the salt logic that configure sd-gpg with the key file, effectively
clearing the path for a potentially disposable sd-gpg qube.

For context, all qubes can make certain qrexec calls to dom0 already
(see qubes.SyncAppMenus, for example).

Fixes #1505

deeplow committed:
Make sd-gpg load dom0-provided key(s) on boot

Replace sd-gpg SaltStack configuration with qubes-services that replace
its behavior. The logic has been moved to a package on the client repo.

deeplow commented:
@conorsch I decided to take a look at what exists out there. Turns out the answer is "a lot". I've documented these in https://github.com/freedomofpress/securedrop-workstation/issues/1522.

deeplow committed:
Explicitly require dependency 'qubes-gpg-split-dom0'

The dependency was implicit and possibly never noticed as missing given
that it is installed by default in dom0 [1].

[1]: https://github.com/QubesOS/qubes-qubes-release/blob/release4.2/comps/comps-dom0.xml#L1687

deeplow committed:
Add / update gpg tests following on-boot key import

deeplow committed:
Use qube.run() in vm_fingerprint fixture

Unnecessarily long code when we have syntactic sugar for it.

deeplow committed:
Add tests on securedrop.GetJournalistSecretKeys

Tests policy but also server component. The client component is already
indirectly tested through the key being made available in sd-gpg.

RPC test organization:
  - tests/test_vm_sd_gpg.py: includes RPC client tests
  - tests/test_rpc_policy.py: RPC policy test
  - tests/test_rpc_service*.py: RPC server test

legoktm reviewed (changes requested):
I haven't tested this yet but I like where it's going.

deeplow committed:
Re-write GetJournalistSecretKeys in Python

Given the security-critical nature of this script, python is preferable
to make the code clear.

deeplow committed:
Log to stderr in addition to systemd

This makes error messages available to calling qube (useful mostly for
development purposes).

legoktm commented:
(Rebased)

legoktm commented on files/securedrop.GetJournalistSecretKeys:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
I think all those 3 things are doable :) Let me give it a try...

legoktm committed:
Refactor GetJournalistSecretKeys.py

* All caps for global constants, move SECRET_KEYPATH into one
* log -> logger

legoktm reviewed (changes requested):
Functionally this is great. A few remaining things:
* You should review my changes and approve + squash them into yours.
* I flagged some stuff about the tests, once those are fixed I think we can try and OpenQA run now that the securedrop-gpg-config package should be on apt-test.

deeplow committed:
Remove no-longer-needed test ('.profile' reset)

deeplow committed:
Organizes RPC-related files under 'rpc' namespace

Separates policy file tests from particular RPC service tests:
  - tests/test_rpc_policy.py -- policies
  - tests/test_rpc_service_*.py -- custom RPC services

deeplow reviewed (commented):
> You should review my changes and approve + squash them into yours.

I've done this now. The code included some parts that didn't feel relevant given the context, so I refactored it quite a bit. Please take a look.

deeplow committed:
Remove "Journalist" from RPC name (make it shorter)

Just easier to type and reference. Due to the length, some related files
(e.g. gpg-config/securedrop-get-secret-keys.sh) already excluded
"journalist" from the name. And since the only secret keys in the system
belong to journalists, this is kind of implied.

deeplow committed:
Remove leftover 'self' in RPC policy tests

Lingering artifact from a previous refactor

legoktm reviewed (changes requested):
I'm having the same failure as 4.3 OpenQA locally:
```
# test_policy_from_sdgpg_to_dom0_allowed[sd-gpg]
# failure: 

# timestamp None
src_qube_name = 'sd-gpg'

    @pytest.mark.parametrize(
        "src_qube_name",
        [vm.name for vm in Qubes().domains if SD_TAG in vm.tags],
    )
    @pytest.mark.provisioning
    def test_policy_from_sdgpg_to_dom0_allowed(src_qube_name):
        allowed = policy_exists(src_qube_name, "@adminvm", "securedrop.GetSecretKeys")
    
        if src_qube_name == "sd-gpg":
>           assert allowed
E           assert False

tests/test_rpc_policy.py:101: AssertionError
```

deeplow committed:
sd-gpg policy test: iterate over existing VMs

Iterate over existing VMs (sdw_tagged_vms), rather than the expected
list of VMs. This is because there could be a situation where a VM is
missing from the SD_VMS list, meaning that it would not be checked
against. This way we ensure we test against what's actually deployed in
the system. This is consistent with other tests (e.g.
test_sdlog_from_sdw_to_sdlog_allowed).

deeplow commented:
This was in fact happening. I'll through it in my TODO list to investigate why that was happening. In the meantime, replacing `@adminvm` with `dom0` worked just fine.

deeplow committed:
Fix policy checking for securedrop.GetSecretKeys

Apparently Qubes 4.2 and 4.3 work differently when it comes to
translating @adminvm to dom0 and vice versa. I see some code explicitly
doing this [1], so I am not sure it is a bug.

But in the behavior is:
  - Qubes 4.2: dom0 ---translates-to--> @adminvm
  - Qubes 4.3: dom0 ---translates-to--> dom0

Therefore we have to add a switch case to disambiguate. Either that or
wrap around our qrexec_policy() function.

[1]: https://github.com/QubesOS/qubes-core-qrexec/pull/204

legoktm reviewed (approved):
Let's do this!
